doc/developer/filepoints.md

```diff
@@ -0,0 +1,64 @@
+# Introduction
+
+Failpoints are a way to force instrumented code to take a particular action if a given code line is reached while the failpoint is enabled. This is useful when writing tests -- for example, Materialize can be forced to panic or I/O calls can be made to return an error.
+
+Materializes uses the [fail create](https://docs.rs/crate/fail) to enable this functionality.
+
+# Intrumenting the code
+
+## Panic-only failpoints
+
+If a failpoint needs to only cause a panic, the code needs to be instrumented with the following:
+
+```
+fail_point!("read-dir");
+```
+
+## Failpoints that can return an error
+
+If a failpoint needs to be able to cause an error to be returned, the following syntax can be used to supply the error:
+
```

Comment:
If you need the failpoint to return an error, use the following:

(Is "failpoint reached" the error message in this case? It may be too obvious to devs to need stating.)

Reply:
Yes, the example is kind of obvious but you could have a more verbose error message that would allow you to debug your test and make sure you are hitting the exact failpoint you are aiming for.